Configure

```diff
@@ -1484,6 +1485,28 @@ unless ($disabled{afalgeng}) {
 
 push @{$config{openssl_other_defines}}, "OPENSSL_NO_AFALGENG" if ($disabled{afalgeng});
 
+unless ($disabled{ktls}) {
+   $config{ktls}="";
+   if ($target =~ m/^linux/) {
+       my $minver = 4*10000 + 1*1300 + 0;
+       if ($config{cross_compile_prefix} eq "") {
+           my $verstr = `uname -r`;
+           my ($ma, $mi1, $mi2) = split("\\.", $verstr);
+          ($mi2) = $mi2 =~ /(\d+)/;
+          my $ver = $ma*10000 + $mi1*100 + $mi2;
+          if ($ver < $minver) {
+              $disabled{ktls} = "too-old-kernel";
+          }
+      } else {
+          $disabled{ktls} = "cross-compiling";
```

Comment:
Is cross-compilation actually mission impossible? Besides, in the context headers are more reliable source than current uname -r...

Reply:
I didn't try cross-compilation. I guess it might work. It would be better if someone who can actually test it would provide the patches.

Regarding "uname -r", I just copied it from afalg. Please let me know if you have a better solution.
I don't know what you mean by context-headers. Please explain.

Reply:
> context-headers.

No dash. It's supposed to be "in the context [of this request] headers ...". [BTW, should one add comma after context?] Anyway, what I'm trying to say is that I reckon that headers are more reliable basis for taking the decision than uname -r at the moment of config script execution. As simplest example consider scenario when you installed new kernel and headers but haven't rebooted the machine yet. Which would incidentally be equivalent of cross compilation. I mean cross compilation at the very least is nothing else but compilation with alien headers. Well, one can always argue that "nobody does that" or "if they do they are doing it wrong", but it's really just a projection of one's prior experiences, and nothing says that it's the only possible experience, or that it's most satisfying one... Of course it might turn out that what I'm trying to say is not as practical as implied, but it's not inappropriate to keep trying to push boundaries, because if we don't, we won't make as much progress...

Reply:
But on more essential note I'd say that it's more than appropriate that for *starters* it's disabled by default.

Reply:
Well, this is not *formally* right either. Earlier mentioned case of not-rebooted-my-computer-yet was a simplest example of cross-compilation. So simple that it would be totally appropriate to put it in quotes, "cross-compilation". Formally right thing to do is to call *configured* compiler with *configured* flags, and ask it to compile or pre-process something. This is most reliable assumption one can make. Idea is to hold user responsible for providing right target headers for *configured* compiler invoked with *configured* flags.

Reply:
I understand that for cross-compilation the configured compiler,linker,etc. need to be called, but the ktls configuration is not responsible for that.
The only thing that is relevant for this config option is the linux headers. Your previously suggested use-case should work now. But, cross-compilers usually don't provide any special headers, let alone linux kernel headers. Moreover, I haven't seen a relevant OpenSSL tutorial that mentions include path manipulation for cross-compilation.

Reply:
> cross-compilers usually don't provide any special headers,

They totally are. Cross-compiler without target platform headers is worthless.

> let alone linux kernel headers.

Even this is not true. If target platform is Linux, then cross-compiler will have headers that are customarily found in what corresponds to Debian linux-libc-dev.

> I haven't seen a relevant OpenSSL tutorial

If you play your cards right --cross-compiler-prefix is all it takes. For example I cross-compile for ARM Linux with `Configure linux-armv4 --cross-compiler-prefix=arm-linux-gnueabi-` directly on my regular computer. And there is a separate snapshot of glibc and Linux headers [and of course libraries]. Another example is Travis CI cross-compiling for mingw...

Reply:
If you were to add linux kernel headers for the cross-compiled target. Where would you add them relative to your "--cross-compiler-prefix"?

Reply:
Well, it's totally up to cross-compiler provider/vendor, so there is no common answer. But in my case ARM Linux headers can be found in /usr/arm-linux-gnueabi/include/. But once again, beauty is that you don't have to know, just call cross-compiler and pass it to somebody else. Well, unless you are in business of providing cross-compiler, but we aren't and you don't seem to be...

Reply:
Would something like this meet the requirement?


    if ($target =~ m/^linux/) {
        my $usr = "/usr/$config{cross_compile_prefix}";
        chop($usr);
        if ($config{cross_compile_prefix} eq "") {
            $usr = "/usr";
        }
        my $minver = (4 << 16) + (13 << 8) + 0;
        my @verstr = split(" ",`cat $usr/include/linux/version.h | grep LINUX_VERSION_CODE`);
        print STDERR "Linux vesion $verstr[2]. KTLS minmum version $minver.\n";

        if ($verstr[2] < $minver) {
            $disabled{ktls} = "too-old-kernel";
        }
    } else {
        $disabled{ktls}  = "not-linux";
    }